libs/cli/src/main/java/org/elasticsearch/cli/internal/JsonPrintWriter.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cli.internal;
+
+import org.elasticsearch.core.SuppressForbidden;
+
+import java.io.OutputStream;
+import java.io.PrintWriter;
+import java.time.Clock;
+import java.time.Instant;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeFormatterBuilder;
+import java.util.Locale;
+
+/**
+ * {@link PrintWriter} formatting lines as JSON using a subset of fields written by ECSJsonLayout.
+ * Anything that looks like a JSON object is immediately passed through and written as is.
+ */
+public class JsonPrintWriter extends PrintWriter {
+    private static final int INITIAL_BUFFER_SIZE = 256;
+    private static final int MAX_BUFFER_SIZE = 4 * INITIAL_BUFFER_SIZE;
+    private static final DateTimeFormatter ISO_INSTANT_FORMATTER = new DateTimeFormatterBuilder().appendInstant(3).toFormatter(Locale.ROOT);
+
+    private final StringBuilder buffer = new StringBuilder(INITIAL_BUFFER_SIZE);
+    private final Clock clock;
+
+    public JsonPrintWriter(OutputStream out, boolean autoFlush) {
+        this(out, autoFlush, Clock.systemUTC());
+    }
+
+    @SuppressForbidden(reason = "Override PrintWriter to emit Json")
+    protected JsonPrintWriter(OutputStream out, boolean autoFlush, Clock clock) {
+        super(out, autoFlush);
+        this.clock = clock;
+    }
+
+    @Override
+    public void write(String msg) {
```

Comment:
Is this override sufficient? Other methods could be called on this PrintWriter, for example `write(char[])`. And ultimately each message could have newlines.

Taking a step back, could we handle output json at a higher level than the print writer? For example, if we assume that anything grabbing the stderr/stdout knows what they are doing, we could just handle jsonifying in the Terminal for print/println (though we still need to handle buffering there I think).

For passing through stdout/stderr from the ES process, the ErrorPumpThread currently takes in a PrintWriter, but all it uses is println and flush. I wonder if that could be narrowed to a unique interface so that we can handle jsonification separately there since that is the one that we need to look for existing json, while any writing directly to the Terminal in the cli should always output as json when we are in json mode.

Reply:
> Taking a step back, could we handle output json at a higher level than the print writer? For example, if we assume that anything grabbing the stderr/stdout knows what they are doing, we could just handle jsonifying in the Terminal for print/println (though we still need to handle buffering there I think).

Actually, I initially considered doing that, but dropped it realizing `ErrorPumpThread` uses the stderr `PrintWriter`; same for `printStackTrace` in the launcher shutdown hook. However, it would certainly greatly simplify things. `ErrorPumpThread` could of course be rewritten using `Terminal` (optionally with a flag if to use stdout / stderr), though that seems a bit weird. The better alternative would certainly be a narrowed interface as suggested below, but ...

> For passing through stdout/stderr from the ES process, the ErrorPumpThread currently takes in a PrintWriter, but all it uses is println and flush. I wonder if that could be narrowed to a unique interface so that we can handle jsonification separately there since that is the one that we need to look for existing json, while any writing directly to the Terminal in the cli should always output as json when we are in json mode.

A limited closable, flushable `Printer` just supporting `print(String)`, `println(String)` and `println()` was another idea I considered. Turns out most of the use cases of `Terminal.getWriter()` and `Terminal.getErrorWriter()` require an actual `PrintWriter` and those are often out of our control (e.g. jopt, `printStracktrace`). Exposing both, the `PrintWriter`s and a similar narrowed `Printer` supporting JSON output creates a very confusing surface I think. Unfortunately, to me at least, this looks like a dead end.

> Is this override sufficient? Other methods could be called on this PrintWriter, for example write(char[]). And ultimately each message could have newlines.

These are certainly two fair and sketchy points. 

- Initially I only planned to support jsonification for the higher level print[ln] APIs (not sure why I overwrote `write` rather than `print`). Anyways, it's easy to break/circumvent jsonification this way and likely it's not sufficient. And then there's also cases like `write('{')`. Consider all consecutive writes to be json until writing `}` or the next `println()` or only detect JSON if using the print[ln] APIs? 

- Newlines are ambiguous in this context I feel. Currently newlines within a string are escaped (`\` + `n`) for JSON and effectively removed, emitting a newline requires `println`. IMHO this behavior is a closer match to emitting JSON log, but the other alternative is certainly as valid. 

Reflecting on the above I see two options forward:

1) Move jsonification into `Terminal` (maintaining a buffer per stdout/stderr) and let `ErrorPumpThread` forward lines via the `Terminal`. While the simplest approach, there's a rather strange dependency on `Terminal` in `ErrorPumpThread` and any usage of `printStracktrace` won't emit JSON (currently only the shutdownhook). Though, those stacktraces would be scrampled over multiple JSON objects anyways which isn't great either.

2) Overwrite all necessary `Writer` APIs in `PrintWriter`. My proposal here would be to only detect JSON if using `print[ln](String)` (being just us in `ErrorPumpThread`) and consider everything else to not be JSON (meaning it will be buffered & converted). Regarding newlines my preference is to keep the current behavior, but I'm open to changing that.

I'm leaning towards 2), but let me know your thoughts.

Reply:
I've pushed a much more exhaustive / robust impl of the JsonPrintWriter & tests covering all PrintWriter apis. JSON is only passed through untouched if using the print*ln* APIs, everything else will buffer lines to emit JSON on the next `println()`.